Bugfix/channel switch wrong bandwidth

tomereli opened the pull request:
Fix several bugs in the channel selection implementation:
- Fix wrong bandwidth use in DWPAL which caused chan_switch command to fail
- Fix bug in agent's channel selection which selected a restricted channel
- Fix bug in agent's channel selection which caused the agent not to send operating channel report when channel switch was not needed.
- Fix bug in agent's channel selection which did not follow the Multi-AP specification by not marking channel preference with empty channel list as non-operable

WiFi Alliance support case # 00129476 helped identify the final bug. With this, we pass 4.2.1 without any workarounds.

Fixes #344

Commit:
bwl: dwpal: use actual bandwidth in channel_switch

The hostapd CHAN_SWITCH command accepts the following bandwidth values:
20, 40, 80, 160.
Commit 71105d7 Fixed a bug in the BWL nl80211 implementation which used
the eWiFiBandwidth enum value instead of the integer value it represents
for the channel switch command.
The same bug exists in the BWL dwpal implementation, so fix it now.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: refactor channel selection handling

Channel selection request includes a list of operating classes and a
list of restricted channels for each operating class.
The radio agent, upon receipt of the request, needs to check if the
current channel is restricted, and switch to a non restricted channel if
it is.
The current implementation goes through all the supported channels, and
for each one it checks if it is restricted, by searching the list of
restricted channels for that operating class which was received from the
controller.

In case the list is empty for a specific operating class, it means that
there are no restrictions other than to check that the channel we are
checking belongs to that operating class - and here lies the bug - we
are only comparing the operating class when looking for restricted
channels in the lambda function, so when the restricted channel list is
empty, we simply skip it and select the previousely set channel.

We could fix it simply by taking out the test for matching operatinf
class outside of the find_if section, but while we're at it, refactor
the function - it makes more sense to reverse the double for loop so
that the outer loop will go through the channel preference.

New logic - for each channel preference received, search for the first
channel from that operating class which is not restricted.
If found, switch to that channel.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: always send operating channel report

Currently, if following a channel selection request we select a channel
which we are already on, we simply ignore the request.

However, we should send an operating channel report even if we do not
switch channel according to the EasyMesh specification:
"If a Multi-AP Agent sent a Channel Selection Response message indicating acceptance of the Multi-AP Controller’s
request for a given radio, the Multi-AP Agent shall make any necessary adjustments to the operating channel, operating
classes and transmit power of its radios and then, irrespective of whether any adjustments have been made, send an
Operating Channel Report message per section 17.1.13 containing information regarding the current operating
parameters for each of the Multi-AP Agent’s radios."

In case tx power change is requested as part of the channel selection
request and channel switch itself is not required, the agent should
update the tx power and then send an operating channel report with the
new tx power value.

since currently setting tx power is not supported in BWL and therefore
there is no event for the tx power being set, the current implementation
simply never sends the operating channel report, which causes
MAP-4.2-ChanSel-MAUT.txt flow in the UCC certification to fail.

Add a workaround for always sending the operating channel report without
waiting for the tx limit missing event.
This solution indeed violates the specification by reporting a wrong tx
power value, but this is not tested currently in the certification, so
we can fix it later once we support setting tx power in the BWL.

To be completed as part of
https://github.com/prplfoundation/prplMesh/issues/725.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: channel selection split to functions

Split channel selection request handling into functions:
- channel_selection_get_channel_preference
- channel_selection_get_transmit_power_limit
- channel_selection_current_channel_restricted
- channel_selection_select_channel

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

YuriyMasechko commented:
According to design channel selection request should contain only restricted channels. But it still could contain empty list and we should count this. So pls skip operating classes with empty list and don't add  them to preferences.

YuriyMasechko commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4555,9 +4555,13 @@ bool slave_thread::handle_channel_selection_request(Socket *sd, ieee1905_1::Cmdu
-    if (!switch_required && !power_limit_received) {
+    if (!switch_required) {
+        LOG(INFO) << "Channel switch not required, sending operating channel report";
         send_operating_channel_report();
```
Is this temporary solution? If power limit was received, channel report should contain new power limit.

tomereli replied:
Yes this is a temporary solution so we can pass certification. I added a comment and a TODO message linking this to https://github.com/prplfoundation/prplMesh/issues/725

Commit:
agent: channel selection: handle empty channel list

The channel selection request includes a channel preference TLV which
holds operating class lists which marks the controller preference.
Each operating class list includes the list of channels for which this
preference applies.

The current implementation therefore goes through all channels in the
list and takes them into account as non-operable (based on the global
preference).
If the channel list for a specific operating class is empty, but still
reported, all channels in that operating class are ignored, meaning not
restricted if the preference is non-operable for all channels.

However, according to Table 23 in the MultiAP Specification, an empty channel list field
indicates that the indicated preference applies to all channels in the operating class.

Due to this, the current implementation is wrong and leads to failure in
test 4.2.1 since the 5G band happens to select a DFS channel for which
the channel list is empty.

As a result, the agent concludes that the current (DFS) channel is not
restricted, and concludes that no channel switch is required, although
the actual preference is to restrict all channels *except for channel
36*.

The solution implemented in this commit is simple - if the current
channel belongs to a restricted operating class (empty channel list),
mark the current channel as restricted.

Not that the current implementation does not take into account any other
preference other than "Non-operable channel" preference, which is
probably wrong, but this can be handled in a future PR.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented:
> According to design channel selection request should contain only restricted channels. But it still could contain empty list and we should count this. So pls skip operating classes with empty list and don't add them to preferences.

See https://github.com/prplfoundation/prplMesh/pull/899/commits/566198a812c7d30a359cca5f894a08eb4c3d8e76

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4475,37 +4475,50 @@ bool slave_thread::handle_channel_selection_request(Socket *sd, ieee1905_1::Cmdu
         if (is_restricted != preference.channels.end()) {
+            LOG(INFO) << "Current channel " << int(hostap_cs_params.channel)
+                      << " is restricted, switch required";
             switch_required = 1;
```
Why `switch_required` is defined as `int`? Clearly it should be `bool` and be set to `true` or `false` instead of `1` or `0`.

tomereli replied:
Correct, but it was int to begin with so I didn't want to change that.
If you think it is critical (you didn't mark this as (nitpick), I can add another commit which does that.

tomereli replied:
NVM apparently I already changed this as part of c7b5669

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4475,37 +4475,50 @@ bool slave_thread::handle_channel_selection_request(Socket *sd, ieee1905_1::Cmdu
-                    !hostap_params.supported_channels[i].is_dfs_channel) {
-                    channel_to_switch = hostap_params.supported_channels[i];
-                    break;
+        for (auto preference : channel_preferences) {
```
It needs to be const reference.

tomereli replied:
fixed

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4475,37 +4475,50 @@ bool slave_thread::handle_channel_selection_request(Socket *sd, ieee1905_1::Cmdu
                          });
```
I don't understand why we set `is_restricted` if the channel from the list is equal to the current channel.
AFAIR it is only restricted if the channel preference is 0.
So IMO it should be looking like this:
```cpp
    for (const auto &preference : channel_preferences) {
        // Check if channel list is with preference of restricted channel
        if (preference.preference.preference == 0) {
            continue;
        }

        auto current_channel_is_restricted = preference.channels.end() == 
                std::find_if(preference.channels.begin(), preference.channels.end(),
                [&hostap_cs_params.channel](const beerocks::message::sWifiChannel &wifi_channel) {
	              return hostap_cs_params.channel == wifi_channel.channel;
	        );

        if (current_channel_is_restricted) {
            LOG(INFO) << "Current channel " << int(hostap_cs_params.channel)
                      << " is restricted, switch required";
            switch_required = true;
            break;
        }
    }
```

Please notice that on my code suggestion, I changed a few things:
1. Use const reference on the main for-loop.
2. Add only `hostap_cs_params.channe` as reference to the lambda function. Adding everything as a reference is useless and it is translated to passing everything possible as a reference to the function, so it hurts performance.
3. Added a `break` when `switch_required` is set to `true`. There is no need to continue with furthermore iterations. 

I'm not sure, but it seems to me that it is possible to merge this loop and the one down below.

tomereli replied:
You are again correct. The current code assumes that `preference` is always `0`, meaning always non-operable. The correct solution is to take the actual preference into account but then we will need to add much more code which is not related to this PR and should be handled in #725 - this I prefer not to do as part of a bug fix PR.

Regarding the change it self - yours is wrong, since if channel list is empty, you don't mark *any* channel as non-operable (restricted). 

However thank you for this comment - I fixed everything you requested.

morantr replied:
Just to clarify, this comment was on the upper for-loop (line 4470)and not the one you have changed.

rmelotte replied:
Just a quick note because I've seen it before, this is not correct:
> Adding everything as a reference is useless and it is translated to passing everything possible as a reference to the function"

Only the variables you use in the lambda body will be captured by reference so you don't need to explicitly list them (see [cppreference](https://en.cppreference.com/w/cpp/language/lambda)). 
Of course you can still list them if you prefer but if it's for "performance reasons", then you don't have to. The 2 are equivalent.

tomereli commented:
> Don't you always need to send the operating channel report, regardless of whether the switch is required?

Yes I do, and I updated the PR to do that. With this we pass the test.

morantr reviewed (commented):
I didn't finish, this is what I got so far.

tomereli commented:
[4.2.1_PASS.zip](https://github.com/prplfoundation/prplMesh/files/4281488/4.2.1_PASS.zip)

mariomaz commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4422,29 +4422,37 @@ beerocks::message::sWifiChannel slave_thread::channel_selection_select_channel()
+    for (const auto &preference : channel_preferences) {
+        // for now we handle only non-operable preference
+        // TODO - handle as part of https://github.com/prplfoundation/prplMesh/issues/725
+        if (preference.preference.preference != 0) {
```
Define an enum instead of using a magic number (i.e.: NON_OPERABLE: 0, MOST_PREFERRED: 15)
(nitpick)

tomereli replied:
Done

mariomaz commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4422,29 +4422,37 @@ beerocks::message::sWifiChannel slave_thread::channel_selection_select_channel()
+    for (const auto &preference : channel_preferences) {
+        // for now we handle only non-operable preference
+        // TODO - handle as part of https://github.com/prplfoundation/prplMesh/issues/725
+        if (preference.preference.preference != 0) {
```
Surely there must be a better name for this: `preference.preference.preference`
(nitpick)

morantr replied:
It bothers me too.

tomereli replied:
Not part of this PR, but I can fix it in a separate commit. Will do.

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4382,16 +4382,66 @@ bool slave_thread::handle_channel_preference_query(Socket *sd, ieee1905_1::CmduM
 {
-    const auto mid = cmdu_rx.getMessageId();
-    LOG(DEBUG) << "Received CHANNEL_SELECTION_REQUEST_MESSAGE, mid=" << std::dec << int(mid);
+    for (auto preference : channel_preferences) {
```
const reference

morantr commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4382,16 +4382,66 @@ bool slave_thread::handle_channel_preference_query(Socket *sd, ieee1905_1::CmduM
+
+bool slave_thread::channel_selection_current_channel_restricted()
+{
+    for (auto preference : channel_preferences) {
```
const reference

mariomaz commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4448,6 +4498,12 @@ bool slave_thread::handle_channel_selection_request(Socket *sd, ieee1905_1::Cmdu
+bool slave_thread::channel_selection_get_transmit_power_limit(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                              int &power_limit)
+{
     for (auto tx_power_limit_tlv : cmdu_rx.getClassList<wfa_map::tlvTransmitPowerLimit>()) {
```
You should use 'const auto&'
I think `cmdu_rx` parameter should also be a const reference

tomereli replied:
Done

mariomaz commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4382,16 +4382,66 @@ bool slave_thread::handle_channel_preference_query(Socket *sd, ieee1905_1::CmduM
+{
     channel_preferences.clear();
 
     for (auto channel_preference_tlv : cmdu_rx.getClassList<wfa_map::tlvChannelPreference>()) {
```
You should use 'const auto&'
I think `cmdu_rx` parameter should also be a const reference

mariomaz replied:
This one has been marked as resolved but it is not.
This file deserves a full review regarding the 'const auto&' usage.

Commit:
bcl: beerocks_message_structs: remove sPreference

There is no need to define sPreference in beerocks_message_struct since
it is only used by the agent and not sent in a CMDU, and it also causes
unecessary encapsulation of oper_class, preference and reason variables
when used externally which causes the access to the actual preference
value to use the following code snippet:
auto preference = preference.preference.preference;

This has been noted out in
https://github.com/prplfoundation/prplMesh/pull/899#discussion_r387543541
so fix it now.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: tlvChannelPreference: add ePreference

Preparative commit.
Add ePreference to based on Table 23 in the Multi-AP specification for
mapping out all possible values for bits 4-7 in the channel preference
TLV.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

mariomaz reviewed (approved):
son_slave_thread.cpp deserves a full review regarding the 'const auto&' usage.
